### ISSUE TITLE
Nested models error: " undefined method `reverse' for nil:NilClass "

### ISSUE THREAD

**Issue by alex-v16** [nesquena/query_reviewer#27]
Hello! I use nested models(think reason near this) and hard relations between them. When I try to create some model instance, i get such error in create action.

NoMethodError in AdvertsController#create
undefined method `reverse' for nil:NilClass

When gem is disabled - error disappears.

**Comment by furiabhavesh** [comment]
We are also facing same issue which does not come if we remove this gem.

**Comment by rubenrails** [comment]
Same error here.

**Comment by sudhirj** [comment]
That's a +1. The error happens here: 

https://github.com/rails/rails/blob/master/activerecord/lib/active_record/connection_adapters/abstract/database_statements.rb#L14

Is QR setting `binds` to `nil` or something?

**Comment by jdelStrother** [comment]
I could be wrong, but I believe this is caused by Rails running its built-in EXPLAIN query, and QueryReviewer running an EXPLAIN on that.  I'm not sure it's anything to do with nested models.

[EDIT: nevermind, I just reproduced it without an EXPLAIN.  So much for that theory...]

**Comment by laurens** [comment]
same here

activerecord (3.2.11)

https://github.com/rails/rails/blob/v3.2.11/activerecord/lib/active_record/connection_adapters/abstract/database_statements.rb#L8

**Comment by laurens** [comment]
It works for me when going back to query_reviewer 0.1.6

**Comment by laurens** [comment]
No, unfortunately it does not. But this seems only to happen on POST requests.

**Comment by nesquena** [comment]
Hmm, anyone have a working theory for what's causing this? Unfortunately not using this project at the moment but still maintaining and happy to accept patches.

**Comment by jdelStrother** [comment]
Is everyone here using the mysql2 gem, and only seeing it on inserts?

**Comment by jdelStrother** [comment]
So...   I can reproduce just by loading up Rails console, enabling query reviewer with : 

``` ruby
Thread.current["queries"] = QueryReviewer::SqlQueryCollection.new; Thread.current["query_reviewer_enabled"] = true
```

and then performing an insert (eg `Post.create()`).

The problem seems to result from some state being stored in the mysqladapter itself:

ActiveRecord::ConnectionAdapters::DatabaseStatements#insert calls `DatabaseStatements#to_sql(arel, binds)`.  If you call `arel.to_sql` after this point, (like QueryReviewer does in sanitize_strings_and_numbers_from_sql), then you'll hit the  `undefined method 'reverse' for nil:NilClass` error.  If you call `arel.to_sql` _before_ this point, it works fine.

I think the problem results from the BindSubstitution class, which once you've run `DatabaseStatements#to_sql` stashes an `@block` ivar which changes the behaviour of `to_sql`.  My head is starting to hurt following all the arel visitor methods though, so I'm struggling to pin it down much further.

**Comment by eyaleizenberg** [comment]
This error is still happening on rails 3.2.13 with latest query reviewer.
